### PR TITLE
CFY-2708: Add support for use_external_resource property on the cloudify.vcloud.nodes.Server type.

### PR DESCRIPTION
**Pull request by nmishkin**
This change is modeled on a similar feature in the OpenStack plugin.

### CODE REVIEW

File: server_plugin/server.py [hunk 1]
```diff
@@ -174,16 +185,17 @@ def create(vca_client, **kwargs):
...
+        vapp_name = get_vapp_name(ctx.instance.runtime_properties)
+        config = get_vcloud_config()
+        vdc = vca_client.get_vdc(config['vdc'])
+        vapp = vca_client.get_vapp(vdc, vapp_name)
+        if _vapp_is_on(vapp) is False:
+            ctx.logger.info("Power-on VApp {0}".format(vapp_name))
+            task = vapp.poweron()
+            if not task:
+                raise cfy_exc.NonRecoverableError("Could not power-on vApp")
+                wait_for_task(vca_client, task)
```
Comment (ran-z) [inline server_plugin/server.py]: seems like this has been pushed one indentation too far?

File: server_plugin/server.py [hunk 2]
```diff
@@ -194,6 +206,9 @@ def start(vca_client, **kwargs):
 @operation
 @with_vca_client
 def stop(vca_client, **kwargs):
+    if ctx.node.properties.get('use_external_resource'):
```
Comment (ran-z) [inline server_plugin/server.py]: minor nitpick, i apologize in advance: this style of returning if it is in fact use_external_resource is contradictory to the opposite which is used in the "start" method. personally i'd use the one here for both; i think it'd be a bit more readable if we were to be consistent about it in any case.

File: server_plugin/server.py [hunk 3]
```diff
@@ -208,6 +223,9 @@ def stop(vca_client, **kwargs):
 @operation
 @with_vca_client
 def delete(vca_client, **kwargs):
+    if ctx.node.properties.get('use_external_resource'):
```
Comment (ran-z) [inline server_plugin/server.py:232]: note that even when using an external resource, the "delete" operation should clean up the runtime properties stored on the node instance.

File: server_plugin/server.py [hunk 4]
```diff
@@ -254,7 +272,7 @@ def _get_state(vca_client):
         for connection in nw_connections}
 
     for connection in nw_connections:
-        if connection['network_name'] == management_network_name:
+        if ctx.node.properties.get('use_external_resource') or connection['network_name'] == management_network_name:
```
Comment (ran-z) [inline server_plugin/server.py]: not sure i understand this part. if i use an external resources, why are all nw_connections necessarily the management network?

**Comment by ran-z** [comment]
regarding the use of "ctx.node.properties.get('use_external_resource')" - I'm not sure I understand the part with the "backwards compatibility" 0lvin mentioned here - i think it should actually use [] dictionary retrieval and not .get(), as its a mandatory property - and any legacy code should be updated to suit this new schema.

**Comment by ran-z** [comment]
regarding having the management network as mandatory - it is true that at the moment, that's also the case with the openstack plugin. however this is actually something we want to amend; a VM shouldn't necessarily have to be connected to the management network; it must simply be accessible from the management machine.
i'd make sure to avoid forcing this coupling, or if it's already coupled, at least i'd avoid adding extra layers onto this coupling.

File: server_plugin/server.py (hunk 4 above)
Comment (nmishkin) [inline server_plugin/server.py]: I already planned to remove this change based on the exchange I had with @0lvin . I had originally thought I'd be able to allow the management network to be omitted for external resources and would use the (admittedly questionable) logic I introduced here. I now see that was a mistake and that the specification of the management network will be mandatory for external resources.

**Comment by 0lvin** [comment]
For run our mock tests for your case:
nosetests tests/mock --with-coverage --cover-package=server_plugin --cover-html

And need to add tests for case when we have "use_external_resource" == True and check that your changes work as you predicted like return from function after check that flag. 

I will merge this changes and create example for you case.

**Comment by 0lvin** [comment]
Tests example for this case: dcbd0415e55a5adda5a3c51d72b5433b4415bd25 CFY2708: update server_plugin.server test

**Pull request merged**